Storage/src/StorageObject.php

```diff
@@ -775,177 +832,61 @@ public function downloadAsStreamAsync(array $options = [])
-            if (strpos($value, PHP_EOL) !== false) {
-                throw new \InvalidArgumentException(
-                    'Line endings are not allowed in header values. Replace line endings with a single space.'
-                );
-            }
-
-            // Invalid header names throw exception.
-            if (strpos($name, 'x-goog-') !== 0) {
-                throw new \InvalidArgumentException(
-                    'Header names must begin with `x-goog-`.'
-                );
-            }
-
-            $headers[] = $name .':'. $value;
-        }
+        $rw = $this->connection->requestWrapper();
 
-        if ($headers) {
-            $headers[] = '';
+        $keyFile = $this->pluck('keyFile', $options, false);
+        if ($keyFile) {
+            $scopes = $this->pluck('scopes', $options, false) ?: $rw->scopes();
+            $credentials = CredentialsLoader::makeCredentials($scopes, $keyFile);
+        } else {
+            $credentials = $rw->credentials();
         }
 
-        $objectPieces = explode('/', $this->identity['object']);
-        array_walk($objectPieces, function (&$piece) {
-            $piece = rawurlencode($piece);
-        });
-        $objectName = implode('/', $objectPieces);
-
         $resource = sprintf(
             '/%s/%s',
             $this->identity['bucket'],
-            $objectName
+            $this->identity['object']
         );
 
-        $toSign = [
-            $options['method'],
-            $options['contentMd5'],
-            $options['contentType'],
-            $seconds,
-            implode("\n", $headers) . $resource,
-        ];
-
-        // NOTE: While in most cases `PHP_EOL` is preferable to a system-specific character,
-        // in this case `\n` is required.
-        $string = implode("\n", $toSign);
-        $signature = $this->signString($keyFile['private_key'], $string, $options['forceOpenssl']);
-        $encodedSignature = urlencode(base64_encode($signature));
-
-        $query = [];
-        $query[] = 'GoogleAccessId=' . $keyFile['client_email'];
-        $query[] = 'Expires=' . $seconds;
-        $query[] = 'Signature=' . $encodedSignature;
-
-        if ($options['responseDisposition']) {
-            $query[] = 'response-content-disposition=' . urlencode($options['responseDisposition']);
-        } elseif ($options['saveAsName']) {
-            $query[] = 'response-content-disposition=attachment;filename='
-                . urlencode('"' . $options['saveAsName'] . '"');
-        }
-
-        if ($options['responseType']) {
-            $query[] = 'response-content-type=' . urlencode($options['responseType']);
-        }
-
-        if ($this->identity['generation']) {
-            $query[] = 'generation=' . $this->identity['generation'];
-        }
+        $version = $this->pluck('version', $options, false) ?: self::DEFAULT_URL_SIGNING_VERSION;
```

Comment:
Could you also log a warning that in the future v4 will be the default signing version?
Similar message as above:
> You have generated a signed URL using the default v2 signing implementation. In the future, this will default to v4. You may experience breaking changes if you use longer than 7 day expiration times with v4. To opt-in to the behavior specify config.version='v4'

Reply:
Do you mean a logged notice here?

Reply:
Yes, is that reasonable?

Reply:
yep!

Reply:
https://github.com/googleapis/google-cloud-php/pull/1742/files#diff-57f8ebcee88a0b79aad2f6763d8418d6R115